baker/Deployer.ts

```diff
@@ -0,0 +1,374 @@
+import * as fs from "fs-extra"
+import * as prompts from "prompts"
+import ProgressBar = require("progress")
+import { execWrapper } from "../db/execWrapper"
+import { spawn } from "child_process"
+import simpleGit, { SimpleGit } from "simple-git"
+import { WriteStream } from "tty"
+import { ProgressStream } from "./ProgressStream"
+
+const TEMP_DEPLOY_SCRIPT_SUFFIX = `.tempDeployScript.sh`
+
+enum DeployTarget {
+    staging = "staging",
+    hans = "hans",
+    playfair = "playfair",
+    jefferson = "jefferson",
+    nightingale = "nightingale",
+    explorer = "explorer",
+    exemplars = "exemplars",
+    tufte = "tufte",
+    roser = "roser",
+}
+
+interface DeployerOptions {
+    owidGrapherRootDir: string
+    userRunningTheDeploy: string
+    target: DeployTarget
+    skipChecks?: boolean
+    runChecksRemotely?: boolean
+}
+
+const OWID_STAGING_DROPLET_IP = "165.22.127.239"
+const OWID_LIVE_DROPLET_IP = "209.97.185.49"
+
+export class Deployer {
+    private options: DeployerOptions
+    private progressBar: ProgressBar
+    private stream: ProgressStream
+    constructor(options: DeployerOptions) {
+        this.options = options
+        const { target, skipChecks, runChecksRemotely } = this.options
+
+        this.stream = new ProgressStream(process.stderr)
+        // todo: a smarter way to precompute out the number of steps?
+        const testSteps = !skipChecks && !runChecksRemotely ? 2 : 0
+        this.progressBar = new ProgressBar(
+            `Baking and deploying to ${target} [:bar] :current/:total :elapseds :name\n`,
+            {
+                total: 11 + testSteps,
+                renderThrottle: 0, // print on every tick
+                stream: (this.stream as unknown) as WriteStream,
+            }
+        )
+    }
+
+    private async runAndTick(command: string) {
+        await execWrapper(command)
+        this.progressBar.tick({ name: `✅  finished ${command}` })
+    }
+
+    private get targetIsStaging() {
+        return new Set(Object.values(DeployTarget)).has(this.options.target)
+    }
+
+    get targetIsProd() {
+        return this.options.target.toString() === "live"
+    }
```

Comment:
The type of `this.options.target` is `DeployTarget`, so I guess that shouldn't even allow for `live`?

Reply:
You're right. That was ugly. Just cleaned that up.